core/src/main/java/org/apache/calcite/sql/fun/SqlCastFunction.java

```diff
@@ -93,6 +95,29 @@ public SqlCastFunction() {
 
   //~ Methods ----------------------------------------------------------------
 
+  @Override public void validateCall(final SqlCall call, final SqlValidator validator,
+      final SqlValidatorScope scope,
+      final SqlValidatorScope operandScope) {
+    try {
+      super.validateCall(call, validator, scope, operandScope);
+    } catch (UnsupportedOperationException ex) {
+      assert call.operandCount() == 2;
+      throw validator.newValidationError(call,
+          RESOURCE.cannotCastToType(call.operand(1).toString()));
+    }
+  }
+
+  @Override public RelDataType deriveType(final SqlValidator validator,
+      final SqlValidatorScope scope, final SqlCall call) {
+    try {
+      return super.deriveType(validator, scope, call);
+    } catch (UnsupportedOperationException ex) {
+      assert call.operandCount() == 2;
+      throw validator.newValidationError(call,
+          RESOURCE.cannotCastToType(call.operand(1).toString()));
```

Comment:
Should the original `UnsupportedOperationException` be kept as a `cause` of the exception?

Reply:
You are right, I think it should. 

However, I found that incorporating the cause will make this a much larger change than it is right now. Currently, `SqlValidator` interface only has a `newValidationError` method that takes an `ExInst`, which represents an exception without cause. We can add a new method `newValidationErrorWithCause` to the `SqlValidator` interface. There will need to be some refactoring for `SqlUtil.newContextException` as well. Note that hacking this by adding `initCause` to the resulting exception returned from `newValidationError` does not work because the exception has already been initialized with a cause during construction.

Overall, I think this can be split into a refactoring commit + this commit. I think the combined LOC will be  >100. Not sure if the extra code added and refactoring is worth the benefit: on one hand, having the cause provides better context. On the other hand, the cast function is not that complicated `CAST(value as type)` and the current exception message might be enough for debugging.

What do you think?

Reply:
Then I'm not sure the current patch is a net win. I guess `UnsupportedOperationException` can be thrown for lots of reasons, and producing `cannotCastToType` can be misleading.

It did not expect adding a cause would require that level of refactoring, however, I agree something has to be done with `ExInst` and `ExInstWithCause`.

For instance: `ExInst withCause(Throwable)` could be added to `ExInst`.
An alternative option is probably to remove `ExInstWithCause` altogether.

Reply:
I sent a new pull request at https://github.com/apache/calcite/pull/2326 for this. Should I close this pull request?